packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -708,12 +708,12 @@ export function forceRehydration(vm: VM) {
     }
 }
 
-export function hasShadow(vm: VM): vm is VM & { cmpRoot: ShadowRoot } {
+export function isLightRenderModeVM(vm: VM): vm is VM & { cmpRoot: ShadowRoot } {
     // We don't refer to vm.def.ctor.shadow because that could be changed by user
```

Comment:
this comment is incorrect: `vm.def.ctor.shadow`

any extraction that should happen during the definition, should happen there, and we should never use def.ctor because that one can change, as you pointed out.

Reply:
I believe that for any declaration value that the engine cares about, the `def` should have a boolean value extracted once. In this case, the problem is how do you determine if `vm.cmpRoot` should be created or not. I will like to get that changed now as well.

Reply:
Addressing this in a different PR.